hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -679,18 +682,130 @@ BlobProperty getBlobProperty(Path blobPath,
    * {@link AbfsClient#getBlobProperty(Path, TracingContext)} call
    */
-  BlobProperty getContainerProperty(TracingContext tracingContext) throws AzureBlobFileSystemException {
-    AbfsRestOperation op = client.getContainerProperty(tracingContext);
-    BlobProperty blobProperty = new BlobProperty();
+  BlobProperty getContainerProperty(TracingContext tracingContext)
+      throws AzureBlobFileSystemException {
+    try (AbfsPerfInfo perfInfo = startTracking("getContainerProperty", "getContainerProperty")) {
+      LOG.debug("getContainerProperty for filesystem: {} path: {}",
+          client.getFileSystem());
 
-    final AbfsHttpOperation opResult = op.getResult();
+      AbfsRestOperation op = client.getContainerProperty(tracingContext);
+      perfInfo.registerResult(op.getResult()).registerSuccess(true);
 
-    blobProperty.setIsDirectory(true);
-    blobProperty.setPath(new Path("/"));
+      BlobProperty blobProperty = new BlobProperty();
+      blobProperty.setIsDirectory(true);
+      blobProperty.setPath(new Path(FORWARD_SLASH));
 
-    return blobProperty;
+      return blobProperty;
+    }
+  }
+
+  /**
+   * Gets user-defined properties(metadata) of the blob over blob endpoint.
+   * @param path
+   * @param tracingContext
+   * @return hashmap containing key value pairs for blob metadata
+   * @throws AzureBlobFileSystemException
+   */
+  public Hashtable<String, String> getBlobMetadata(final Path path,
+      TracingContext tracingContext) throws AzureBlobFileSystemException {
+    try (AbfsPerfInfo perfInfo = startTracking("getBlobMetadata", "getBlobMetadata")) {
+      LOG.debug("getBlobMetadata for filesystem: {} path: {}",
+          client.getFileSystem(),
+          path);
+
+      final AbfsRestOperation op = client.getBlobMetadata(path, tracingContext);
+      perfInfo.registerResult(op.getResult()).registerSuccess(true);
+
+      final Hashtable<String, String> metadata = parseResponseHeadersToHashTable(op.getResult());
+      return metadata;
+    }
+    catch (AbfsRestOperationException ex) {
+      // The path does not exist explicitly.
+      // Check here if the path is an implicit dir
+      if (ex.getStatusCode() == HttpURLConnection.HTTP_NOT_FOUND && !path.isRoot()) {
+        List<BlobProperty> blobProperties = getListBlobs(path, null,
+            tracingContext, 2, true);
+        if (blobProperties.size() == 0) {
+          throw ex;
+        }
+        else {
+          // Path exists as implicit directory.
+          // Return empty hashmap for properties
+          return new Hashtable<>();
+        }
+      }
+      else {
+        throw ex;
+      }
+    }
+  }
+
+  /**
+   * Sets user-defined properties(metadata) of the blob over blob endpoint.
+   * @param path on which metadata is to be set
+   * @param metadata set of user-defined properties to be set
+   * @param tracingContext
+   * @throws AzureBlobFileSystemException
+   */
+  public void setBlobMetadata(final Path path,
```

Comment:
What about implicit directory?

Reply:
This call will fail for implicit direcotry with 404. This behavior is at par with DFS endpoint (Current go live drop)

Reply:
have a concern there:

dfs doesn't understand explicit/implicit, for implicit its nothing in there.

if dev want to assign some property to a directory, they shouldn't care if its explicit or implicit. if its a valid path in tree (virtually), we should honor setXAttr. What you say?

Reply:
closing comment: keeping behavior with dfs endpoint.